File: counter.c
// Description of the Program

/*The given C code presents a straightforward multi-threaded program that lets mutexes (an internal locking mechanism of an operating system) synchronize access to a shared resource that in this case is a counter variable. The program makes use of two scripts, namely thread1 and thread2, each one of which increases the value stored in a shared counter variable. The mutex operation is used to protect the counter from the multiple thread access by allowing only one thread to modify the counter at a time and implying a race conditions prevention. Also, mutex1 should be used. Interestingly, the conditional statement in thread1 changes the counter by plus 100 while mutex2 protects the operation. The program makes the previous feature possible by storing the information on how many times thread1 gets the bonus increment (100) through a separate variable called bonusCount.
*/

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

#define MAX_COUNT 2000000

int counter = 0;
int bonusCount = 0;
pthread_mutex_t mutex;

void *thread1_func(void *arg) {
    // Entry section
    for (int i = 0; i < MAX_COUNT; i++) {
        pthread_mutex_lock(&mutex); // Critical section entry
        if (counter % 100 == 0) {
            counter += 100;
            bonusCount++; // Increment bonus count
        } else {
            counter++;
        }
        pthread_mutex_unlock(&mutex); // Critical section exit
    }
    return NULL;
}

void *thread2_func(void *arg) {
    // Entry section
    for (int i = 0; i < MAX_COUNT; i++) {
        pthread_mutex_lock(&mutex); // Critical section entry
        counter++;
        pthread_mutex_unlock(&mutex); // Critical section exit
    }
    return NULL;
}

int main() {
    pthread_t thread1, thread2;

    // Initialize the mutex
    pthread_mutex_init(&mutex, NULL);

    // Create threads
    pthread_create(&thread1, NULL, thread1_func, NULL);
    pthread_create(&thread2, NULL, thread2_func, NULL);

    // Wait for threads to finish
    pthread_join(thread1, NULL);
    pthread_join(thread2, NULL);

    // Remainder section
    printf("Final counter value: %d\n", counter);
    printf("Thread1 bonus count: %d\n", bonusCount);
    printf("Thread2 did %d updates, counter = %d\n", MAX_COUNT, counter);

    // Destroy the mutex
    pthread_mutex_destroy(&mutex);

    return 0;
}

File: Counter.java
// Description of the Program
/*
The following Java code is a sample of a multi-threaded program that uses a ReentrantLock to synchronize the use of a bounded resource, like for example a counter variable. The program uses two threads, thread1 and thread2, that add to the modifiable shared counter variable. The counter is guarded by a ReentrantLock to make only a single thread change its value at a time and thus eliminate the race conditions. Moreover, although thread1 typically updates the counter by 1, it occasionally increments it by 100 which is ensured by the respective lock. In addition, the program maintains that how frequently thread1 is titled the bonus increment is counted by incrementing the separate bonusCount variable.
*/

public class Counter {
    private static int value = 0;
    private static final java.util.concurrent.locks.ReentrantLock mutex = new java.util.concurrent.locks.ReentrantLock();
    private static int bonusCount = 0;

    public static void main(String[] args) throws InterruptedException {
    // Entry section
    Thread thread1 = new Thread(() -> {
        for (int i = 0; i < 2000000; i++) {
            mutex.lock(); // Critical section entry
            try {
                if (value % 100 == 0) {
                    value += 100;
                    bonusCount++; // Increment bonus count
                } else {
                    value++;
                }
            } finally {
                mutex.unlock(); // Critical section exit
            }
        }
    });

    Thread thread2 = new Thread(() -> {
        for (int i = 0; i < 2000000; i++) {
            mutex.lock(); // Critical section entry
            try {
                value++;
            } finally {
                mutex.unlock(); // Critical section exit
            }
        }
    });

    thread1.start();
    thread2.start();

    thread1.join(); // Wait for Thread1 to finish
    thread2.join(); // Wait for Thread2 to finish

    // Remainder section
    System.out.printf("Final counter value: %d\n", value);
    System.out.printf("Thread1 bonus count: %d\n", bonusCount);
    System.out.printf("Thread2 did %d updates, counter = %d\n", 2000000, value);
}

}
